<a href="https://colab.research.google.com/github/jagvgithub/Algortimos_optimizacion_VIU/blob/main/Seminario_Algoritmos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Algoritmos de optimización - Seminario<br>
Nombre y Apellidos: Jose Antonio Gómez Vargas

Url: https://github.com/jagvgithub/Algortimos_optimizacion_VIU/blob/main/Seminario_Algoritmos.ipynb<br>

Problema 3. Combinar cifras y operaciones

Descripción del problema:

* El problema consiste en analizar el siguiente problema y diseñar un algoritmo que lo resuelva.
* Disponemos de las 9 cifras del 1 al 9 (excluimos el cero) y de los 4 signos básicos de las
operaciones fundamentales: suma(+), resta(-), multiplicación(*) y división(/)
* Debemos combinarlos alternativamente sin repetir ninguno de ellos para obtener una
cantidad dada. Un ejemplo sería para obtener el 4:
4+2-6/3*1 = 4

(*) La respuesta es obligatoria





                                        

In [ ]:
# librerías necesarias
import pandas as pd
from itertools import permutations

# Divide y vencerás

In [ ]:
# creación de funciones necesarias dentro del algoritmo

def generar_expresiones(numeros, operadores):
    expresiones = []
    resultados = []
    for num_perm in permutations(numeros, 5):
        for op_perm in permutations(operadores, 4):
            expresion = "{}{}{}{}{}{}{}{}{}".format(num_perm[0], op_perm[0], num_perm[1], op_perm[1], num_perm[2], op_perm[2], num_perm[3], op_perm[3], num_perm[4])
            expresiones.append(expresion)
    return expresiones

def evaluar_expresiones(expresiones):
    resultados = []
    for expresion in expresiones:
        try:
            resultado = eval(expresion)
            resultados.append(resultado)
        except ZeroDivisionError:
            pass
    return resultados

# Validación del valor de entrada
while True:
    try:
        valor_objetivo = int(input("\nPor favor ingrese el valor deseado como resultado en la expresión: "))
        if valor_objetivo <= 0:
            print("\nEl número de discos debe ser mayor que cero, Inténtalo de nuevo.")
        else:
            break
    except ValueError:
        print("\nPor favor, ingresa un número entero.")

# Listas de números y operadores
numeros = [1, 2, 3, 4, 5, 6, 7, 8, 9]
operadores = ['+', '-', '*', '/']

# Generar todas las permutaciones posibles de 5 números y 4 operadores
expresiones = generar_expresiones(numeros, operadores)

# Evaluar los resultados de las expresiones
resultados = evaluar_expresiones(expresiones)

# Filtrar las expresiones que cumplen con el valor objetivo
expresiones_resultantes = [expresiones[i] for i in range(len(expresiones)) if resultados[i] == valor_objetivo]

# Crear un DataFrame con las expresiones y los resultados
df = pd.DataFrame({'Expresión': expresiones_resultantes, 'Resultado': [valor_objetivo] * len(expresiones_resultantes)})

df= df[df['Resultado'] == valor_objetivo]

# Mostrar el DataFrame
print(df)



Por favor ingrese el valor deseado como resultado en la expresión: 4
Valor mínimo: 4
Valor máximo: 4
      Expresión  Resultado
0     1-2*3/6+4          4
1     1-2/3*6+7          4
2     1/2*4-3+5          4
3     1/2*4+5-3          4
4     1/2*4-5+7          4
...         ...        ...
2107  9-8+6*1/2          4
2108  9-8+6/2*1          4
2109  9-8+6*2/4          4
2110  9-8+6/4*2          4
2111  9-8*6/4+7          4

[2112 rows x 2 columns]


(*)¿Cuantas posibilidades hay sin tener en cuenta las restricciones?<br>

¿Cuantas posibilidades hay teniendo en cuenta todas las restricciones.

In [ ]:
''' Pregunta 1 '''

numeros = len(set(numeros))  # Número de opciones para cada posición del número
operadores = len(set(operadores))  # Número de opciones para cada posición del operador
posibilidades_numeros = numeros ** 5
posibilidades_operadores = operadores ** 4
total_posibilidades = posibilidades_numeros * posibilidades_operadores
print("Total de posibles soluciones sin restricciones: {:,}".format(total_posibilidades))

''' Pregunta 2 '''

numeros = 9
signos = 4
# Calcular el número total de permutaciones de 5 números y 4 signos sin repetición
num_permutaciones_numeros = len(list(permutations(range(1, numeros + 1), 5)))
num_permutaciones_signos = len(list(permutations(['+', '-', '*', '/'], signos)))
total_posibilidades = num_permutaciones_numeros * num_permutaciones_signos
print("Total de posibles soluciones con restricciones en numeros y signos: {:,}".format(total_posibilidades))


print("Total de posibles soluciones con restricciones en numeros y signos y el valor objetivo del usuario: {:,}".format(len(df)))

df.head(10)



Total de posibles soluciones sin restricciones: 15,116,544
Total de posibles soluciones con restricciones en numeros y signos: 362,880
Total de posibles soluciones con restricciones en numeros y signos y el valor objetivo del usuario: 2,112


Expresión  Resultado
0  1-2*3/6+4          4
1  1-2/3*6+7          4
2  1/2*4-3+5          4
3  1/2*4+5-3          4
4  1/2*4-5+7          4
5  1*2+4-6/3          4
6  1/2*4-6+8          4
7  1/2*4+7-5          4
8  1/2*4-7+9          4
9  1/2*4+8-6          4

Modelo para el espacio de soluciones<br>
(*) ¿Cual es la estructura de datos que mejor se adapta al problema? Argumentalo.(Es posible que hayas elegido una al principio y veas la necesidad de cambiar, arguentalo)


Para determinar cuál es la mejor estructura de datos para este problema, consideremos los requisitos y las operaciones que necesitamos realizar:

*Almacenamiento y Acceso Secuencial:* Necesitamos almacenar todas las posibles expresiones y sus resultados. El acceso secuencial es necesario para evaluar cada expresión y comparar el resultado con el valor objetivo.

*Flexibilidad en el Tamaño:* Dado que el número de combinaciones posibles puede
ser muy grande, necesitamos una estructura de datos que pueda crecer dinámicamente.

* Listas

Las listas son una buena opción para este problema porque cumplen con los requisitos mencionados:

*Acceso Secuencial:* Nos permiten iterar fácilmente sobre las expresiones y resultados.
*Almacenamiento Dinámico:* Podemos agregar elementos sin necesidad de definir su tamaño previamente.

*Flexibilidad:* Manejan cualquier número de elementos, lo cual es útil para un gran número de combinaciones.
Además, las listas en Python tienen un buen rendimiento para la mayoría de las operaciones básicas (como agregar elementos y acceder a ellos secuencialmente), lo cual es suficiente para este problema.

---
Diferencias entre Fuerza Bruta y Divide y Vencerás

---


- Fuerza Bruta

Método: Prueba todas las combinaciones posibles sin ninguna optimización.

Ventajas: Simple de implementar.

Desventajas: Muy ineficiente para problemas de gran tamaño debido al alto número de combinaciones posibles.

- Divide y Vencerás

Método: Divide el problema en subproblemas más pequeños, resuelve cada uno independientemente y luego combina las soluciones.

Ventajas: Puede reducir significativamente el tiempo de ejecución al simplificar el problema y permitir soluciones más rápidas a subproblemas más pequeños.

Desventajas: Puede ser más complejo de implementar.

Según el modelo para el espacio de soluciones<br>
(*)¿Cual es la función objetivo?



El ejercicio proporciona un valor objetivo que representa el resultado deseado que una expresión matemática debe alcanzar. Este valor objetivo se utiliza como criterio de filtrado para seleccionar solo aquellas expresiones que produzcan el valor deseado. En este contexto, la "función objetivo" se refiere a este valor deseado ingresado como entrada, que las expresiones generadas deben alcanzar. No se trata de una función matemática en el sentido tradicional, sino más bien de un objetivo que se persigue al generar y evaluar las expresiones matemáticas.

(*)¿Es un problema de maximización o minimización?

Este problema no implica la optimización de una función objetivo para encontrar el máximo o el mínimo valor. En lugar de eso, el objetivo es encontrar todas las combinaciones posibles de números y operadores cuyo resultado, al ser evaluado, coincida con un valor particular proporcionado. Por lo tanto, no puede clasificarse estrictamente como un problema de maximización o minimización, ya que no estamos buscando optimizar una función objetivo en términos de su valor máximo o mínimo. En su lugar, estamos explorando todas las combinaciones para encontrar aquellas que satisfagan el criterio establecido.

Diseña un algoritmo para resolver el problema por fuerza bruta

# Fuerza bruta

In [ ]:
# Listas de números y operadores
numeros = [1, 2, 3, 4, 5, 6, 7, 8, 9]
operadores = ['+', '-', '*', '/']

# Generar todas las permutaciones posibles de 5 números y 4 operadores
expresiones = []
resultados = []
for num_perm in permutations(numeros, 5):
    for op_perm in permutations(operadores, 4):
        expresion = "{}{}{}{}{}{}{}{}{}".format(num_perm[0], op_perm[0], num_perm[1], op_perm[1], num_perm[2], op_perm[2], num_perm[3], op_perm[3], num_perm[4])
        try:
            resultado = eval(expresion)
            expresiones.append(expresion)
            resultados.append(resultado)
        except ZeroDivisionError:
            pass

# Crear un DataFrame con las expresiones y los resultados
df = pd.DataFrame({'Expresión': expresiones, 'Resultado': resultados})

minimo_fuerza_bruta = df['Resultado'].min()
maximo_fuerza_bruta = df['Resultado'].max()

filtered_df = df[df['Resultado'].isin([1, 2, 3, 4, 5, 6, 7, 8, 9])]


# creación de df independientes

# Crear un diccionario para almacenar los DataFrames independientes para cada valor único en 'Resultado'
dfs_por_resultado = {}

# Iterar sobre los valores únicos en 'Resultado'
for resultado in filtered_df['Resultado'].unique():
    # Filtrar el DataFrame original para el valor único en 'Resultado'
    df_resultado = filtered_df[filtered_df['Resultado'] == resultado]
    # Asignar el DataFrame filtrado al diccionario
    dfs_por_resultado[f'df_{int(resultado)}'] = df_resultado


df = df[df['Resultado'] == valor_objetivo]

# Mostrar el DataFrame
print(df)

        Expresión  Resultado
249     1-2*3/6+4        4.0
299     1-2/3*6+7        4.0
743     1/2*4-3+5        4.0
862     1/2*4+5-3        4.0
911     1/2*4-5+7        4.0
...           ...        ...
361446  9-8+6*1/2        4.0
361567  9-8+6/2*1        4.0
361614  9-8+6*2/4        4.0
361831  9-8+6/4*2        4.0
361905  9-8*6/4+7        4.0

[2112 rows x 2 columns]


- ¿Qué valor máximo y mínimo se pueden obtener según las condiciones del problema?

- ¿Es posible encontrar todos los valores enteros posibles entre dicho mínimo y máximo ?

In [ ]:
print("Valor mínimo:", minimo_fuerza_bruta)
print("Valor máximo:", maximo_fuerza_bruta)

if minimo_fuerza_bruta < 1 and maximo_fuerza_bruta > 9:
    print("Es posible encontrar todos los valores enteros posibles entre dicho mínimo y máximo")
else:
    print("No es posible encontrar todos los valores enteros posibles entre dicho mínimo y máximo")

Valor mínimo: -70.71428571428571
Valor máximo: 78.83333333333333
Es posible encontrar todos los valores enteros posibles entre dicho mínimo y máximo


Calcula la complejidad del algoritmo por fuerza bruta

Para analizar su complejidad, consideramos los siguientes pasos::

1. **Generar Permutaciones de Números**:
    - Total de permutaciones posibles de 5 números seleccionados de una lista de 9:
      $$ P(9, 5) = \frac{9!}{(9-5)!} = 15120 $$

2. **Generar Permutaciones de Operadores**:
    - Total de permutaciones posibles de 4 operadores seleccionados de una lista de 4:
      $$ P(4, 4) = 4! = 24 $$

3. **Evaluar Expresiones**:
    - Total de expresiones generadas y evaluadas:
      $$ P(9, 5) \times P(4, 4) = 15120 \times 24 = 362880 $$

### Conclusión

La complejidad total del algoritmo es:

$$ O(\text{total}) = O(P(9, 5)) \times O(P(4, 4)) = O(15120) \times O(24) = O(362880) $$

Esto se simplifica a:

$$ O(\text{total}) = O(n!) $$

donde \( n \) es el número de elementos de los cuales se toman las permutaciones. En este caso específico:

$$ O(9!) = O(362880) $$

En resumen, la complejidad del algoritmo de fuerza bruta es factorial en la cantidad de números permutados, lo que lo hace altamente ineficiente para grandes conjuntos de entrada.


(*)Diseña un algoritmo que mejore la complejidad del algortimo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta

In [ ]:
# Acceder a los DataFrames independientes por su nombre

df_1 = dfs_por_resultado['df_1']
df_2 = dfs_por_resultado['df_2']
df_3 = dfs_por_resultado['df_3']
df_4 = dfs_por_resultado['df_4']
df_5 = dfs_por_resultado['df_5']
df_6 = dfs_por_resultado['df_6']
df_7 = dfs_por_resultado['df_7']
df_8 = dfs_por_resultado['df_8']
df_9 = dfs_por_resultado['df_9']


filtered_df['Resultado'].value_counts().sort_index()

Resultado
1.0    2008
2.0    1688
3.0    1952
4.0    2112
5.0    2880
6.0    2200
7.0    2824
8.0    2376
9.0    2584
Name: count, dtype: int64

(*)Calcula la complejidad del algoritmo

Estos números representan el número de expresiones válidas para cada resultado entero del 1 al 9.

Si utilizas estos números para calcular la complejidad del algoritmo mejorado, puedes estimar la cantidad total de operaciones necesarias para generar todas las expresiones válidas para cada resultado y luego combinarlas en los DataFrames independientes.

Por ejemplo, si sumamos estos números, obtenemos el total de expresiones válidas para todos los resultados:

$$2008+1688+1952+2112+2880+2200+2824+2376+2584=20624$$

Por lo tanto, con esta mejora, solo necesitarías evaluar un total de 20624 expresiones en lugar de las $9!=362880$ expresiones del enfoque de fuerza bruta original.

Esta mejora en la complejidad es significativa y puede hacer que el algoritmo sea mucho más eficiente en términos de tiempo de ejecución y recursos computacionales requeridos. La reducción en el número de operaciones necesarias se traduce en un mejor rendimiento y una menor carga computacional.

Según el problema (y tenga sentido), diseña un juego de datos de entrada aleatorios

In [ ]:
# Pedir al usuario que ingrese el valor

valor = input("Por favor ingrese un valor del 1 al 9: ")

# Verificar si el valor ingresado está en el rango válido
if valor.isdigit() and 1 <= int(valor) <= 9:
    # Obtener el nombre del DataFrame correspondiente al valor ingresado
    nombre_df = f'df_{valor}'
    # Acceder al DataFrame usando el nombre
    if nombre_df in globals():
        df = globals()[nombre_df]
        # Mostrar el DataFrame
        print(df)
    else:
        print(f"No existe un DataFrame para el valor {valor}.")
else:
    print("El valor ingresado no es válido.")



Por favor ingrese un valor del 1 al 9: 7
        Expresión  Resultado
297     1-2*3/6+7        7.0
751     1-2+4/3*6        7.0
815     1/2*4-3+8        7.0
962     1+2*4-6/3        7.0
966     1-2+4*6/3        7.0
...           ...        ...
361672  9*8/6+2-7        7.0
361715  9-8/6*3+2        7.0
361753  9+8-6/3*5        7.0
361968  9+8-6*5/3        7.0
362081  9*8/6-7+2        7.0

[2824 rows x 2 columns]


Aplica el algoritmo al juego de datos generado




Enumera las referencias que has utilizado(si ha sido necesario) para llevar a cabo el trabajo

Respuesta

Describe brevemente las lineas de como crees que es posible avanzar en el estudio del problema. Ten en cuenta incluso posibles variaciones del problema y/o variaciones al alza del tamaño

Para avanzar en el estudio del problema y mejorar el enfoque de fuerza bruta, se pueden considerar varias estrategias:

* Optimización del algoritmo de fuerza bruta: Aunque el enfoque de fuerza bruta es simple, se puede mejorar optimizando el código para reducir el tiempo de ejecución. Por ejemplo, se pueden evitar cálculos redundantes o implementar técnicas de poda para eliminar ramas innecesarias en la búsqueda de soluciones.

* Implementación de algoritmos más eficientes: En lugar de generar todas las permutaciones posibles y evaluar cada una, se pueden explorar algoritmos más eficientes, como la programación dinámica o el uso de estructuras de datos especializadas, que pueden reducir significativamente el tiempo de ejecución.

* Consideración de variaciones del problema: El problema actual implica encontrar expresiones que den como resultado números enteros específicos. Se puede considerar la posibilidad de extender el problema para permitir números no enteros o para buscar expresiones que se aproximen a un valor objetivo en lugar de ser exactamente iguales.

* Escalabilidad del algoritmo: Si se necesita manejar tamaños de problema más grandes, es crucial asegurarse de que el algoritmo pueda escalar eficientemente. Esto implica diseñar estructuras de datos y algoritmos que sean eficientes incluso para tamaños de entrada grandes.

* Exploración de paralelismo y distribución: Para acelerar el procesamiento en sistemas con múltiples núcleos o en entornos distribuidos, se puede explorar la posibilidad de paralelizar el algoritmo o distribuir la carga de trabajo entre varios dispositivos.